lib/profileAgents.js

```diff
@@ -0,0 +1,361 @@
+  const paKeystore = await kms.createKeystore(
+    {capabilityAgent: paZcapAgent, referenceId: 'primary'});
+  const paKeystoreAgent = kms.getKeystoreAgent(
+    {capabilityAgent: paZcapAgent, keystore: paKeystore});
+
+  const did = key.id.split('#')[0];
+  const profileAgent = {
+    id: did,
+    sequence: 0,
+    account: accountId,
+    controller: {
+      id: capabilityAgent.id,
+      seed: secret,
+      keystore: keystore.id
+    },
+    keystore: paKeystore.id,
+    capabilityInvocationKey: (await key.getKeyDescription()).id
+  };
+  const now = Date.now();
+  const meta = {created: now, updated: now};
+  let record = {
+    meta,
+    profileAgent
+  };
+  try {
+    const collection = utils.getCollection('profile-profileAgent');
+    const result = await collection.insert(record, database.writeOptions);
+    record = result.ops[0];
+  } catch(e) {
+    if(!database.isDuplicateError(e)) {
+      throw e;
+    }
+    throw new BedrockError(
+      'Duplicate profile agent.',
+      'DuplicateError', {
+        public: true,
+        httpStatusCode: 409
+      }, e);
+  }
+  return {
+    ...record,
+    profileAgent: {
+      ...record.profileAgent,
+      capabilityAgent: paZcapAgent,
+      keystoreAgent: paKeystoreAgent
+    }
+  };
+};
+
+/**
+ * Get a Profile Agent
+ *
+ * @param {String} id the id of the ProfileAgent.
+ *
+ * @return {Promise<ProfileAgent>} resolves to a ProfileAgent.
+ */
+exports.get = async ({id} = {}) => {
+  assert.string(id, 'id');
+
+  const query = {'profileAgent.id': id};
+  const collection = utils.getCollection('profile-profileAgent');
+  const record = await collection.findOne(query, {_id: 0});
+  if(!record) {
+    const details = {
+      profileAgent: id,
+      httpStatusCode: 404,
+      public: true
+    };
+    throw new BedrockError(
+      'Profile agent not found.',
+      'NotFoundError', details);
+  }
+  const {capabilityAgent, keystoreAgent} = await _getAgents(record);
+  return {
+    ...record,
+    profileAgent: {
+      ...record.profileAgent,
+      capabilityAgent,
+      keystoreAgent
+    }
+  };
+};
+
+/**
+ * Get all Profile Agent(s)
+ *
+ * @param {String} accountId the id of the account associated with the
+ *                           ProfileAgent(s).
+ *
+ * @return {Promise<Arrray<ProfileAgent>>} resolves to a ProfileAgent(s).
+ */
+exports.getAll = async ({accountId} = {}) => {
+  assert.string(accountId, 'accountId');
+
+  const query = {'profileAgent.account': accountId};
+  const collection = utils.getCollection('profile-profileAgent');
+  const records = await collection.find(query, {_id: 0}).toArray();
+  if(records.length === 0) {
+    return [];
+  }
+  // TODO: Find proper promise-fun library for concurrency
+  const promises = records.map(async record => {
+    const {capabilityAgent, keystoreAgent} = await _getAgents(record);
+    return {
+      ...record,
+      profileAgent: {
+        ...record.profileAgent,
+        capabilityAgent,
+        keystoreAgent
```

Comment:
Yeah, the more I see this pattern, the more I think we're doing too much. Let's expose `getAgents` and let the caller use that if they want on the `record`.